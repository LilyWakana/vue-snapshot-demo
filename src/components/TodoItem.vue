<template>
  <li
    :class="['todo-item', item.finished?'todo-finished':'']"
    @mouseover="handleItemMouseIn"
    @mouseleave="handleItemMouseLeave"
  >
    <input type="checkbox" v-model="item.finished">
    <span class="content">{{item.content}}</span>
    <button class="del-btn" v-show="!item.finished&&hover" @click="emitDelete">delete</button>
  </li>
</template>

<script>
export default {
  name: "TodoItem",
  props: {
    item: Object
  },
  data() {
    return {
      hover: false
    };
  },
  methods: {
    handleItemMouseIn() {
      this.hover = true;
    },
    handleItemMouseLeave() {
      // mouseleave 不冒泡
      // mouseout 冒泡，所以在子元素上也会触发
      this.hover = false;
    },
    emitDelete() {
      this.$emit("delete");
    }
  }
};
</script>

<style lang="scss">
.todo-item {
  list-style: none;
  padding: 4px 16px;
  height: 22px;
  line-height: 22px;
  .content {
    margin-left: 16px;
  }
  .del-btn {
    margin-left: 16px;
  }
  &.todo-finished {
    text-decoration: line-through;
  }
}
</style>
