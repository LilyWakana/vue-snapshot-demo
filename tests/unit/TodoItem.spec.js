import TodoItem from '@/components/TodoItem.vue';
import {
    shallowMount
} from '@vue/test-utils';
import {
    renderToString
} from '@vue/server-test-utils';
import {
    createRenderer
} from 'vue-server-renderer';

describe('TodoItem test', () => {
    it('finished render', () => {
        const wrapper = shallowMount(TodoItem, {
            propsData: {
                item: {
                    finished: true,
                    content: 'test TodoItem'
                }
            }
        });
        const checkbox = wrapper.find('input'),
            span = wrapper.find('span'),
            button = wrapper.find('button');
        expect(checkbox.element.checked).toBe(true)
        expect(span.text()).toBe('test TodoItem')
        expect(button.isVisible()).toBe(false)
    })

    it('unfinished render', () => {
        const wrapper = shallowMount(TodoItem, {
            propsData: {
                item: {
                    finished: false,
                    content: 'play basketball'
                }
            }
        });
        const checkbox = wrapper.find('input'),
            span = wrapper.find('span'),
            button = wrapper.find('button');
        expect(checkbox.element.checked).toBe(false);
        expect(span.text()).toBe('play basketball');
        expect(button.isVisible()).toBe(false);

    })

    it('mouseover event', () => {
        const wrapper = shallowMount(TodoItem, {
            propsData: {
                item: {
                    finished: false,
                    content: 'test TodoItem'
                }
            }
        })
        const button = wrapper.find('button');
        wrapper.trigger("mouseover")
        expect(button.isVisible()).toBe(true)
    })

    it('mouseleave event', () => {
        const wrapper = shallowMount(TodoItem, {
            propsData: {
                item: {
                    finished: false,
                    content: 'test TodoItem'
                }
            }
        })
        const button = wrapper.find('button');
        wrapper.trigger("mouseover")
        wrapper.trigger('mouseleave')
        expect(button.isVisible()).toBe(false)

    })
})

describe('TodoItem snapshot test', () => {
    it('first render', () => {
        const wrapper = shallowMount(TodoItem, {
            propsData: {
                item: {
                    finished: true,
                    content: 'test TodoItem content change'
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('toggle checked', () => {
        const renderer = createRenderer();
        const wrapper = shallowMount(TodoItem, {
            propsData: {
                item: {
                    finished: true,
                    content: 'test TodoItem'
                }
            }
        })
        const checkbox = wrapper.find('input');
        checkbox.trigger('click');
        renderer.renderToString(wrapper.vm, (err, str) => {
            expect(str).toMatchSnapshot()
        })

    })
    
    it('mouseover', () => {
        const renderer = createRenderer();
        const wrapper = shallowMount(TodoItem, {
            propsData: {
                item: {
                    finished: false,
                    content: 'test TodoItem'
                }
            }
        })
        wrapper.trigger('mouseover');
        renderer.renderToString(wrapper.vm, (err, str) => {
            expect(str).toMatchSnapshot()
        })
    })
})