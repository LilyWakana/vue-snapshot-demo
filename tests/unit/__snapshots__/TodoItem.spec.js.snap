// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem snapshot test first render 1`] = `<li class="todo-item todo-finished"><input type="checkbox"> <span class="content">test TodoItem</span> <button class="del-btn" style="display: none;">delete</button></li>`;

exports[`TodoItem snapshot test mouseover 1`] = `<li class="todo-item"><input type="checkbox"> <span class="content">test TodoItem</span> <button class="del-btn">delete</button></li>`;

exports[`TodoItem snapshot test toggle checked 1`] = `<li class="todo-item"><input type="checkbox"> <span class="content">test TodoItem</span> <button class="del-btn" style="display:none;">delete</button></li>`;
